### PR TITLE
Filter and sort clause field name validation

### LINKED ISSUE #690: Filter and sort clause field name validation
Add a filter and sort clause field name validation to make sure `$` operators can not be used. Below queries should error.

`{
  "find": {
      "sort" : {"$gt" : 1}
  }
}`

and 
`{
  "find": {
      "filter" : {"$gt" : {"test" : 5}}
  }
}`

### PR DESCRIPTION
**Pull request by Hazel-Datastax**
**What this PR does**:
Add a filter and sort clause field name validation to make sure $ operators can not be used.

**Which issue(s) this PR fixes**:
Fixes #690 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializer.java [hunk 1]
```diff
@@ -103,18 +103,29 @@ public SortClause deserialize(JsonParser parser, DeserializationContext ctxt)
         sortExpressions.add(exp);
         break;
       } else {
+        if (path.isBlank()) {
+          throw new JsonApiException(
+              ErrorCode.INVALID_SORT_CLAUSE_PATH,
+              String.format(
+                  "%s: sort clause path must be represented as not-blank strings.",
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializer.java:110]: nit: `strings` -> `string`. Also no trailing comma for message.

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/DocumentConstants.java [hunk 2]
```diff
@@ -34,6 +34,9 @@ interface Fields {
     // characters, not length limits (nor empty nor "too long" allowed but validated
     // separately)
     Pattern VALID_NAME_PATTERN = Pattern.compile("[a-zA-Z0-9_\\-]*");
+
+    /** Field name pattern in sort and filter clause. */
+    Pattern VALID_FIELD_NAME_PATTERN = Pattern.compile("[a-zA-Z0-9_.\\-]*");
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/config/constants/DocumentConstants.java:39]: Let's rename this as something like `VALID_PATH_PATTERN`; existing `VALID_NAME_PATTERN` is for names.

File: src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java [hunk 3]
```diff
@@ -902,5 +902,22 @@ public void multipleIdFilterAndOr() throws Exception {
                         "Should only have one _id filter, document id cannot be restricted by more than one relation if it includes an Equal");
               });
     }
+
+    @Test
+    public void wrongPathName() throws Exception {
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java:907]: nit: `wrong` -> `invalid`

File: src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializerTest.java [hunk 4]
```diff
@@ -258,5 +258,15 @@ public void mustNotContainEmptyString() {
 
       assertThat(throwable).isInstanceOf(JsonApiException.class);
     }
+
+    @Test
+    public void wrongPathName() {
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializerTest.java:263]: Same here, `wrong` -> `ivalid`

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java [hunk 5]
```diff
@@ -1599,6 +1599,73 @@ public void OrWithIdIn() {
           .body("data.documents", hasSize(3));
     }
 
+    @Test
+    public void dollarOperatorInSortFieldName() {
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java:1603]: FieldName -> PathExpression

**Review by tatu-at-datastax — COMMENTED** [review]
Looks pretty good, just needs some minor renaming.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ComparisonExpression.java [hunk 6]
```diff
@@ -52,6 +53,13 @@ public ComparisonExpression(
    * @return {@link ComparisonExpression} with equal operator
    */
   public static ComparisonExpression eq(String path, Object value) {
+    if (!DocumentConstants.Fields.VALID_FIELD_NAME_PATTERN.matcher(path).matches()) {
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/ComparisonExpression.java:56]: The constructor for ComparisionExpression is public and is not checked for valid pattern. We do support `$vector` as field name in filter with `$exists` operator.

**Review by maheshrajamani — DISMISSED** [review]

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java [hunk 7]
```diff
@@ -111,6 +111,15 @@ private void populateExpression(
...
     } else {
+      // the key should match pattern or equal to $vector
+      if (!(DocumentConstants.Fields.VALID_PATH_PATTERN.matcher(entry.getKey()).matches()
+          || entry.getKey().equals("$vector"))) {
+        throw new JsonApiException(
+            ErrorCode.INVALID_FILTER_EXPRESSION,
+            String.format(
+                "%s: filter clause path ('%s') contains character(s) not allowed",
+                ErrorCode.INVALID_FILTER_EXPRESSION.getMessage(), entry.getKey()));
+      }
```
Comment (Hazel-Datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:122]: @maheshrajamani The `else` block here is for `$eq` missing, like `{ "find": { "filter" : {"name" : "1"} }}`. Should I allow `$vector` here? If it is, the case `{ "find": { "filter" : {"$vector" :  true}} }` which is not valid but will return empty data `{
    "data": {
        "documents": [],
        "nextPageState": null
    }
}`

Comment (maheshrajamani) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:122]: No, only `$exists` supported for `$vector` in filter

Comment (Hazel-Datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:121]: Solved, could you help me review it again? Thank you!

**Review by maheshrajamani — APPROVED** [review]

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializer.java [hunk 8]
```diff
@@ -103,18 +103,29 @@ public SortClause deserialize(JsonParser parser, DeserializationContext ctxt)
         sortExpressions.add(exp);
         break;
       } else {
+        if (path.isBlank()) {
+          throw new JsonApiException(
+              ErrorCode.INVALID_SORT_CLAUSE_PATH,
+              String.format(
+                  "%s: sort clause path must be represented as not-blank string",
```
Comment (vkarpov15) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/SortClauseDeserializer.java:110]: Minor suggestion: I think "sort clause path cannot be a blank string" is easier to read

**Pull request merged**